source/_posts/2018-01-02-Getting-started-with-Vapor-3.md

```diff
@@ -0,0 +1,344 @@
+
+- Configuration: How do we setup providers and middlewares and how do we configure these?
+- Models: How do we create a model and how do we query it?
+- Routing: How do we setup routes for our project and how does these interface with our controllers?
+- Views: How do we setup views for our project?
+- Commands: How do we setup and run commands?
+
+Besides looking into the above areas using a new Vapor 3 project, we will also look into what it will require to migrate an existing project to Vapor 3.
+
+## Getting started with Vapor 3
+
+Please note that this post is based on a beta version of the framework and things might change before we see the official release of the next major version of Vapor.
+
+### Setting up a project
+
+First, make sure you have the recent version of the Vapor toolbox:
+
+```bash
+brew upgrade vapor
+```
+
+And if you don't have the toolbox already installed, then run:
+
+```bash
+brew install vapor/tap/vapor
+```
+
+Also make sure you're running Swift 4 as this is now required in Vapor 3:
+
+```bash
+swift --version
+```
+
+With that in place, we can now create our Vapor 3 project. For this project, we're going to use the [`api-template`](https://github.com/vapor/api-template/tree/beta) since that has been updated for Vapor 3 (using the `beta` branch):
+
+```bash
+vapor new vapor-3-test --template=api --branch=beta
+```
+
+As always, let's make the Vapor toolbox generate our Xcode project for us:
+
+```bash
+cd vapor-3-test
+vapor xcode
+```
+
+When it's done generating the project you should be able to run the project and when visiting `http://localhost:8080/hello` in your browser you should see the `Hello, world!` page.
+
+When looking through the files in our new project, it's already worth noticing some differences. First up, we don't have `Droplet+Setup.swift` and `Config+Setup.swift`. They are instead replaced by:
+
+- `boot.swift`: The `boot` function is called after initialization but before it starts running, meaning that it is an ideal place for running code that should run each time the application starts. Speaking of "application", notice how the signature now calls the `Droplet` an `Application` now. The signature also implies now that there's only going to be one `Application`. 
+- `configure.swift`: This is where you would configure and setup your application. You can basically think of this as a replacement for your old `Config+Setup.swift` file (although the way you setup is now different).
+- `routes.swift`: This is where your "main" route collection is and this where you would add individual routes or other route collections.
+
+### Configuration
+
+Let's go ahead and try and configure something. For this project, we'll use MySQL as this seems to be very common in Vapor projects. Let's add the MySQL package to our `Package.swift` file:
+
+```swift
+.package(url: "https://github.com/vapor/mysql-driver.git", .branch("beta")),
```

Comment:
This tag does not work currently. Once the package is updated it should be `.exact("3.0.0-beta.1")`. Also it might be renamed to `fluent-mysql` because that's the name of the package.